cardano-constitution/src/Cardano/Constitution/Config/Types.hs

```diff
@@ -0,0 +1,62 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# LANGUAGE TupleSections     #-}
+{-# LANGUAGE TypeApplications  #-}
+module Cardano.Constitution.Config.Types
+    ( ConstitutionConfig(..)
+    , ParamConfig(..)
+    ) where
+
+import Cardano.Constitution.Predicate
+import Cardano.Constitution.ProtocolParam
+import PlutusTx as Tx (makeLift)
+import PlutusTx.AssocMap qualified as Tx
+
+import Data.Aeson as Aeson
+import Data.Aeson.Key qualified as Aeson (toText)
+import Data.Aeson.KeyMap qualified as Aeson (toAscList)
+import Data.Map qualified as M
+import Data.Traversable
+import Language.Haskell.TH.Syntax as TH (Lift)
+
+newtype ConstitutionConfig = ConstitutionConfig
+    { unConstitutionConfig :: Tx.Map (ParamName, ParamId) ParamConfig
+    }
+    deriving stock (TH.Lift)
+    deriving newtype (Eq, Show)
+
+newtype ParamConfig = ParamRails { unParamConfig :: Tx.Map PredName PredValue }
+    deriving stock (TH.Lift)
+    deriving newtype (Eq, Show)
+
+instance FromJSON ConstitutionConfig where
+    parseJSON = withObject "ConstitutionConfig" $ \km ->
```

Comment:
I don't know how better to do it so as to have a PlutusTx.Map underneath (instead of a Haskell record). If you know some better way, let's discuss it!

Reply:
> I don't know how better to do it so as to have a PlutusTx.Map underneath (instead of a Haskell record). If you know some better way, let's discuss it!

Would it be possible to decode it as a list of whatever using some existing decoder and then simply convert it to a map? So that the decoding logic is implicit, only the conversion is explicit.

Not sure if it's a good idea though. Making decoding logic implicit is error-prone if the underlying representation gets changed. Although if you hardcode a specific type to decode to (that is some generic list of pairs of whatever), then conversion will fail upon amendment of `ConstitutionConfig`, so we get the best of both worlds: boilerplate is implicit, important stuff is explicit. Will need a Note then.

Anyway, not too important.